### ISSUE TITLE
Rule proposal: `no-instanceof-builtin-object`

### ISSUE THREAD

**Issue by zanminkian** [sindresorhus/eslint-plugin-unicorn#2452]
### Description

`instanceof` operator in JavaScript has some [limitations](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof#instanceof_and_multiple_realms).

To detect the type of a variable, we should encourage user to use: 
- `typeof` operator
```js
typeof foo === 'string' // detect whether foo is string
```
- `node:util/types` module (for Node)
```js
import {isNumberObject, isMap} from 'node:util/types';
isNumberObject(new Number(123)) // true
isNumberObject(123) // false
isMap(new Map()) // true
```
- `@sindresorhus/is` module (for browser)
```js
import is from '@sindresorhus/is';
is(new Map()) === 'Map' // true
```

Therefore, builtin objects like `Map`/`Promise`/`Error` can be detected by `typeof`, `node:util/types` and `@sindresorhus/is`. Should not use `instanceof`.

Rule `no-instanceof-array` can be removed if this rule is implemented.

### Fail

```js
foo instanceof Array
foo instanceof Map
foo instanceof Promise
```

### Pass

```js
import types from 'node:util/types'
Array.isArray(foo)
types.isMap(foo)
types.isPromise(foo)
```

```js
import is from '@sindresorhus/is'
is(foo) === 'Array'
is(foo) === 'Map'
is(foo) === 'Promise'
```

### Proposed rule name

no-instanceof-builtin-object
